src/module.c

```diff
@@ -1447,6 +1451,46 @@ int populateArgsStructure(struct redisCommandArg *args) {
     return count;
 }
 
+/* RedisModule_AddACLCategory can be used to add new ACL command categories. Category names
+ * can only contain alphanumeric characters, underscores, or dashes. Categories can only be added
+ * during the RedisModule_OnLoad function. Once a category has been added, it can not be removed. 
+ * Any module can register a command to any added categories using RedisModule_SetCommandACLCategories.
+ * 
+ * Returns:
+ * - REDISMODULE_OK on successfully adding the new ACL category. 
+ * - REDISMODULE_ERR on failure.
+ * 
+ * On error the errno is set to:
+ * - EINVAL if the name contains invalid characters.
+ * - EBUSY if the category name is already in use.
+ * - ENOMEM if the number of categories reached the max limit of 64 categories.
+ */
+int RM_AddACLCategory(RedisModuleCtx *ctx, const char *name) {
+    if (!ctx->module->onload) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    if (moduleVerifyResourceName(name) == REDISMODULE_ERR) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    if (ACLGetCommandCategoryFlagByName(name)) {
+        errno = EBUSY;
+        return REDISMODULE_ERR;
+    }
+
+    if (ACLAddCommandCategory(name, 0) == C_OK) {
+        ctx->module->num_acl_categories_added++;
+        return REDISMODULE_OK;
+    } else {
+        ACLCleanupAddedCommandCategories(ctx->module->num_acl_categories_added);
```

Comment:
> if we do, then we need to handle other cases of failure during module loading.

Unless it's prohibitively complex, I think we should be trying to clean up partially loaded state as much as possible. I think it's much easier to reason about, if there are other places we aren't fully handling the failed load we should fix it.

Reply:
I agree @madolson, we should be cleaning up the partially loaded module which fail midway. Here in this case it is very much necessary where a module loads multiple categories.

Reply:
ok, so please make sure to handle all module initialization failures in some generic way.
and please add some comments in ACLCleanupAddedCommandCategories about it's purpose.

Reply:
I have renamed the function to `ACLCleanupCategoriesOnFailure` and added comments to make the purpose clear.
I have also modified the functionality a bit by moving this call to the onload function.
So this will cleanup the categories added by the module, which fails while getting loaded.